src/combinatorics/ramsey/infinite.lean

```diff
@@ -0,0 +1,483 @@
+-/
+structure Inf :=
+(s : set ℕ)
+(pf : infinite s)
+
+instance : has_coe Inf (set ℕ) := ⟨Inf.s⟩
+instance : has_mem ℕ Inf := ⟨λ n H, n ∈ H.s⟩
+instance : has_subset Inf := ⟨λ H₁ H₂, H₁.s ⊆ H₂.s⟩
+
+open color
+
+/--
+Given a function from natural numbers to colors,
+this property describes the numbers which are mapped to red.
+-/
+def reds (f : ℕ → color) (H : set ℕ) := (λ n : ℕ, f n = red ∧ n ∈ H )
+
+/--
+Given a function from natural numbers to colors,
+this property describes the numbers which are mapped to blues.
+-/
+def blues (f : ℕ → color) (H : set ℕ) := (λ n : ℕ, f n = blue ∧ H n)
+
+lemma lt_succ_sum : forall x y : ℕ, x < x + y + 1 :=
+λ x y : ℕ, lt_add_of_pos_right x (succ_pos y)
+
+/--
+Unordered pairs are defined as ordered pairs with the condition
+that the second element is strictly greater than the first.
+-/
+def unordered_pairs (S : set ℕ) :=
+  {p : ℕ × ℕ // p.fst < p.snd ∧ S p.fst ∧ S p.snd}
+
+notation `[ℕ]²` := unordered_pairs univ
+notation `[` S `]²` := unordered_pairs S
+
+/--
+Given a number m and a coloring f, the projection of a coloring
+of unordered_pairs to a coloring of numbers is given by fixing
+the first element as m. Note that red is arbitrarily chosen
+as the color assigned to any number below m.
+-/
+def project (f : [ℕ]² → color) (m : ℕ) : ℕ → color :=
+  λ n : ℕ,
+    dite (m < n)
+      (λ h : m < n, f (subtype.mk (m, n) ⟨h, trivial, trivial⟩))
+      (λ _, red)
+
+/--
+The coloring of an unordered pair {a, b}, where a < b,
+agrees with the projection of a applied to b.
+-/
+lemma project_eq (f : [ℕ]² → color) (p : [ℕ]²):
+  f p = project f p.val.fst p.val.snd :=
+begin
+  unfold project,
+  simp [p.property.left]
+end
+
+/--
+Given coloring of natural numbers, a homogeneous set is
+one which is always mapped to the same color.
+-/
+def homogeneous (f : ℕ → color) (H : Inf) :=
+  ∃ c : color, ∀ n, n ∈ H → f n = c
+
+/--
+Given a coloring f, a homogeneous projection is a set of natural numbers
+together with a single natural number.
+The intended use is a set which is homogeneous when projected
+with the given number and coloring.
+-/
+structure homogeneous_proj (f : [ℕ]² → color) extends Inf := (pt : ℕ)
+
+instance (f : [ℕ]² → color) : has_coe (homogeneous_proj f) Inf := ⟨λ c, c.to_Inf⟩
+
+/--
+A homogeneous projection p is refined by another homogeneous projection q if:
+the number given by q is greater than the number given by p,
+the number given by q is an elment of the set given by p,
+the set given by q is contained in the set given by p,
+and the projection using the number given by q is homogeneous
+on the set given by q.
+-/
+def refines {f : [ℕ]² → color} (p q: homogeneous_proj f) :=
+p.pt < q.pt ∧ q.pt ∈ p.s ∧ q.s ⊆ p.s ∧ homogeneous (project f q.pt) q
+
+infix `<<`:50 := refines
+
+local attribute [instance] prop_decidable
+
+/--
+A version of the infinite pigeon hole principle, tailored to our use case.
+-/
+lemma pigeon_hole_principle (f : ℕ → color) (H : Inf):
```

Comment:
Is there already a proof of PHP in mathlib? If so, can you use it to condense your proof?

Reply:
I looked, but couldn't find it. There can be multiple ways of phrasing the PHP, I myself have proved a couple other versions in another repo, even one that is true classically. This particular version fits exactly the needs of the proofs. But I would be happy to include the PHP elsewhere (and show equivalences, etc) if there is a good spot for it.

for fun, my constructive version is here: https://github.com/JaredCorduan/ramsey/blob/lean-3.4.1/src/classical/pigeon.lean